### ISSUE TITLE
limits processor was not enforced without service restart

### ISSUE THREAD

**Issue by igarny** [perfsonar/pscheduler#270]
Hi,
As with my previous environment #269 and the results https://drive.google.com/open?id=0BwOma9C4hovwTk5paDR0WWVRSWM

You will notice that at some point of time the tests start to succeed. The first successful test is triggered by hand and the results do represent the configured limits: test-16.json

The second test, which comes a minute later, based on the schedule defined (see below), is suddenly a successful one: test-17.json
But there is a problem: the limits are not re-enforced. Meaning the results are quite below the observed results of test-16.json
The problem remains with 19, 21, 23, 25, 29 tests.

<h3>demo test specification</h3>

```
2017-01-23T13:05:14 Debug signal ignored; already not debugging
2017-01-23T13:05:14 Debug discontinued
2017-01-23T13:05:14 Assistance is from localhost
2017-01-23T13:05:14 Converting to spec via https://localhost/pscheduler/tests/throughput/spec
Submitting task...
2017-01-23T13:05:14 Fetching participant list from https://localhost/pscheduler/tests/throughput/participants
2017-01-23T13:05:14 {"dest": "psc1", "source": "psc2", "duration": "PT10S", "schema": 1}
2017-01-23T13:05:15 https://localhost/pscheduler/tests/throughput/participants, status 200
2017-01-23T13:05:15 Got participants: {u'participants': [u'psc2', u'psc1']}
2017-01-23T13:05:15 Lead is psc2, null reason is No reason provided.  (Server is running old software.)
2017-01-23T13:05:15 Pinging psc2
2017-01-23T13:05:15 psc2 is up
2017-01-23T13:05:15 Posting task to https://psc2/pscheduler/tasks
2017-01-23T13:05:15 Data is {"test": {"type": "throughput", "spec": {"dest": "psc1", "source": "psc2", "duration": "PT10S", "schema": 1}}, "schedule": {"repeat": "PT2M", "until": "2017-01-23T13:30"}, "schema": 1}
Task URL:
https://psc2/pscheduler/tasks/aadfbc1a-2226-49db-bb06-8c2387b1b411
2017-01-23T13:05:23 Posted https://psc2/pscheduler/tasks/aadfbc1a-2226-49db-bb06-8c2387b1b411
Fetching first run...
2017-01-23T13:05:23 Fetching https://psc2/pscheduler/tasks/aadfbc1a-2226-49db-bb06-8c2387b1b411/runs/first
2017-01-23T13:05:24 Handing off: pscheduler watch --format text/plain --debug https://psc2/pscheduler/tasks/aadfbc1a-2226-49db-bb06-8c2387b1b411
2017-01-23T13:05:24 Debug signal ignored; already not debugging
2017-01-23T13:05:24 Debug discontinued
2017-01-23T13:05:24 Fetching https://psc2/pscheduler/tasks/aadfbc1a-2226-49db-bb06-8c2387b1b411
```

**Comment by mfeit-internet2** [comment]
Limits are applied when a test is put on the schedule, so if it's already on the schedule when the limits are changed, it's still going to run.

There are complications:  If there's a run between A and B and B changes its limits, A needs to try and reschedule the run.  Alternately, B could delete runs that no longer meet the limits and A would just fail.

This is tangentially related to #148.

**Comment by igarny** [comment]
The way I read you, we are also facing another risk. A common case is to schedule reoccurring tests with no end date through meshconfig. Even if there is some end point in time to which meshconfig is appointing the measurements there are risks, that the test following the change will bypass the limits.
As I understand you are working on it....

**Comment by mfeit-internet2** [comment]
If the limits change in a way that would prevent a repeating task from running, the worst case is that scheduled tasks will continue under the old limit for the duration of the scheduling horizon.  That value is currently 24 hours and will become configurable from the outside in a later release.  Passing the run through the limit processor one last time before actually running the test would close that loophole.  Rescheduling what's already there has some logistical problems on non-lead participants.

Meshconfig doesn't actually schedule never-ending tests, opting instead to create new ones every so often as a way to prevent having cruft accumulate when a system goes away.

Bottom line is that there are some things we can do to tighten the process up, and we probably will.  We're still far ahead of where BWCTL's limit system was, and I think that's a good thing.
